### PR TITLE
Fix the kl_div docs

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67443

Fixes https://github.com/pytorch/pytorch/issues/57459

After discussing the linked issue, we resolved that `F.kl_div` computes
the right thing as to be consistent with the rest of the losses in
PyTorch.

To avoid any confusion, these docs add a note discussing how the PyTorch
implementation differs from the mathematical definition and the reasons
for doing so.

These docs also add an example that may further help understanding the
intended use of this loss.

cc @brianjo @mruberry

Differential Revision: [D32136888](https://our.internmc.facebook.com/intern/diff/D32136888)

### CODE REVIEW

**Review by pmeier — CHANGES_REQUESTED** [review]
Some style suggestions and fixes inline and one larger possible mixup. Let's check this carefully to not mess this up again.

**Review by pmeier — APPROVED** [review]
Only nitpicks and style fixes left.

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -373,77 +373,85 @@ def forward(self, input: Tensor, target: Tensor, var: Tensor) -> Tensor:
 
 
 class KLDivLoss(_Loss):
-    r"""The Kullback-Leibler divergence loss measure
+    r"""The Kullback-Leibler divergence loss.
 
-    `Kullback-Leibler divergence`_ is a useful distance measure for continuous
-    distributions and is often useful when performing direct regression over
-    the space of (discretely sampled) continuous output distributions.
+    For tensors of the same shape :math:`y_{\text{model}},\ y_{\text{obs}}`,
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:378]: The math looks correct to me and I appreciate the rigor you approached this with!

Only thing that concerns me is that it's inconsistent with other losses, which generally use the notation `x_n` for input and `y_n` for target. I think it's okay that it's different, but I have a couple suggestions:
* I'd make the connection between the two variables and input / target very explicit (something like `"where y_{model} is the ``input`` and y_{obs} is the target"`)
* Possibly use `y_{true}` / `y_{pred}` instead? this matches TF and I believe fits well with how ML practitioners think (in terms of predicted values vs. ground truth)

Overall I think it's much clearer what's going on :)

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Added a few comments below

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -373,77 +373,85 @@ def forward(self, input: Tensor, target: Tensor, var: Tensor) -> Tensor:
...
 
-    .. math::
-        \ell(x, y) = \begin{cases}
-            \operatorname{mean}(L), & \text{if reduction} = \text{`mean';} \\
-            \operatorname{sum}(L),  & \text{if reduction} = \text{`sum'.}
-        \end{cases}
+        if not log_target: # default
+            loss_pointwise = target * (target.log() - input)
+        else:
+            loss_pointwise = target.exp() * (target - input)
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torch/nn/modules/loss.py:399]: I feel like Marios version with the default behavior at the top is more readable. In the actual implementation, I would also go for your implementation, but for docs I would value readability higher. Of course this is also subjective.

File: torch/nn/modules/loss.py [hunk 3]
```diff
@@ -373,77 +373,85 @@ def forward(self, input: Tensor, target: Tensor, var: Tensor) -> Tensor:
...
+        As all the other losses in PyTorch, this function expects the first argument,
+        :attr:`input`, to be the output of the model (e.g. the neural network)
+        and the second, :attr:`target`, to be the obsevations in the dataset.
+        This differs from the standard mathematical notation :math:`KL(P\ ||\ Q)` where
+        :math:`P` denotes the distribution of the observations and :math:`Q` denotes the model.
+
+    .. note::
+        :attr:`reduction`\ `= "mean"` doesn't return the true kl divergence value, please use
+        :attr:`reduction`\ `= "batchmean"` which aligns with the mathematical definition.
+        In a future release, `"mean"` will be changed to be the same as `"batchmean"`.
```
[the remark this answers is not in the document]
Comment (BenoitDalFerro) [reply torch/nn/modules/loss.py:424]: @jbschlosser, yes absolutely

File: torch/nn/modules/loss.py (hunk 1 above)
Comment (BenoitDalFerro) [reply torch/nn/modules/loss.py:378]: @jbschlosser, suggest actual explicit reference to the P and Q arguments and the implementation being swapped wrt. to math because this is machine learning framework tradition (cf. Tensorflow)

File: torch/nn/modules/loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/nn/modules/loss.py:399]: I'm with @pmeier on this one.

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -373,77 +373,85 @@ def forward(self, input: Tensor, target: Tensor, var: Tensor) -> Tensor:
...
 
-    This criterion expects a `target` `Tensor` of the same size as the
-    `input` `Tensor`.
+        L(y_{\text{model}},\ y_{\text{obs}})
+            = y_{\text{obs}} \cdot \log \frac{y_{\text{obs}}}{y_{\text{model}}}
+            = y_{\text{obs}} \cdot (\log y_{\text{obs}} - \log y_{\text{model}})
 
-    The unreduced (i.e. with :attr:`reduction` set to ``'none'``) loss can be described as:
+    As to avoid underflow issues, to compute this quantity, this loss expects the argument
+    :attr:`input`---:math:`y_{\text{model}}` in the notation above---in the log-space.
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/nn/modules/loss.py:388]: I'm a big fan of em and en dashes, but I completely see your point. Reverting to parens.

File: torch/nn/modules/loss.py (hunk 1 above)
Comment (Lezcano) [reply torch/nn/modules/loss.py:378]: The obs / model came from the probabilistic take on the KL divergence, but I see that using more layman terms might very well be helpful to many.

Also, using this we get rid of parens / em-dashes :D

**Comment by Lezcano** [comment]
Thanks for the detailed review @jbschlosser ! I just addressed the comments. Could you have a second look at this, please?

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the in-depth improvements :)

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32136888).
